src/vs/editor/contrib/linesOperations/linesOperations.ts

```diff
@@ -672,146 +679,132 @@ export class JoinLinesAction extends EditorAction {
 		});
 	}
 
-	public run(_accessor: ServicesAccessor, editor: ICodeEditor): void {
-		let selections = editor.getSelections();
-		if (selections === null) {
-			return;
+	private joinLines(line: number, model: ITextModel, editor: ICodeEditor): IIdentifiedSingleEditOperation | null {
+		if (line >= model.getLineCount()) {
+			return null;
 		}
-
-		let primaryCursor = editor.getSelection();
-		if (primaryCursor === null) {
-			return;
+		const whitespaceAtEnd = model.getLineContent(line).match(/\s*$/);
+		if (!whitespaceAtEnd) {
+			return null;
 		}
+		const range = new Range(line, model.getLineMaxColumn(line) - whitespaceAtEnd[0].length, line + 1, model.getLineFirstNonWhitespaceColumn(line + 1));
+		if (model.getLineContent(line).length > 0) {
+			return EditOperation.replace(range, ' ');
+		} else {
+			return EditOperation.replace(range, '');
+		}
+	}
 
-		selections.sort(Range.compareRangesUsingStarts);
-		let reducedSelections: Selection[] = [];
-
-		let lastSelection = selections.reduce((previousValue, currentValue) => {
-			if (previousValue.isEmpty()) {
-				if (previousValue.endLineNumber === currentValue.startLineNumber) {
-					if (primaryCursor!.equalsSelection(previousValue)) {
-						primaryCursor = currentValue;
-					}
-					return currentValue;
-				}
+	private _getLinesToJoin(editor: IActiveCodeEditor): IJoinLinesOperation[] {
+		// Construct Join operations
+		let operations: IJoinLinesOperation[] = editor.getSelections().map((s) => {
 
-				if (currentValue.startLineNumber > previousValue.endLineNumber + 1) {
-					reducedSelections.push(previousValue);
-					return currentValue;
-				} else {
-					return new Selection(previousValue.startLineNumber, previousValue.startColumn, currentValue.endLineNumber, currentValue.endColumn);
-				}
-			} else {
-				if (currentValue.startLineNumber > previousValue.endLineNumber) {
-					reducedSelections.push(previousValue);
-					return currentValue;
-				} else {
-					return new Selection(previousValue.startLineNumber, previousValue.startColumn, currentValue.endLineNumber, currentValue.endColumn);
-				}
+			let endLineNumber = s.endLineNumber;
+			if (s.startLineNumber < s.endLineNumber && s.endColumn === 1) {
```

Comment:
What is the significance of the magic number 1?  Could this is improved by adding a name constant to reveal what 1 means?